### DISCUSSION TITLE
Adding to capf in `use-package` sets buffer-local value; how to configure properly?

### DISCUSSION THREAD

**Discussion by garyo** [minad/cape#119] [Q&A] [answered]
I'm following the instructions in the readme to set up cape, like this:
```lisp
;; Additional capf completion sources
(use-package cape
  :init
  ;; Note: order matters here. First one returning a result wins.
  ;; There's also a buffer-local list, which takes precedence over these, as long
  ;; as it ends with ~t~.
  (add-to-list 'completion-at-point-functions #'dabbrev-capf)
  (add-to-list 'completion-at-point-functions #'cape-file)
  (add-to-list 'completion-at-point-functions #'cape-history)
  (add-to-list 'completion-at-point-functions #'cape-keyword)
  (message (format "Loading my capf extensions: %s" completion-at-point-functions))
  (message (format "Global (default) value: %s" (default-value #'completion-at-point-functions)))
  )
```
but I can see from those messages that my `add-to-list` forms are setting the buffer-local value of `completion-at-point-functions`. Note that I'm using `straight.el` and emacs 30 built from yesterday's source. Is there a "standard" way to make sure I'm setting the global value of capf?

In my *Messages* buffer I see this:
```
Loading my capf extensions: (cape-keyword cape-history cape-file dabbrev-capf pcomplete-completions-at-point t)
Global (default) value: (tags-completion-at-point-function)
```

**Comment by garyo** [comment]
Note that this works, but seems complicated:
```emacs-lisp
;; Additional capf completion sources
(use-package cape
  :init
  ;; Note: order matters here. First one returning a result wins.
  ;; Have to set the default (global) value since capf is automatically buffer-local
  ;; when set.
  ;; The buffer-local value, which takes precedence over these, calls these as long
  ;; as it ends with ~t~.
  (let ((capf-sources '(cape-history cape-file cape-keyword cape-dabbrev)))
    (setq-default completion-at-point-functions
                  (append capf-sources
                          (default-value 'completion-at-point-functions))))
  (message (format "Loading my capf extensions: %s" completion-at-point-functions))
  (message (format "Global (default) value: %s" (default-value #'completion-at-point-functions)))
  )
```

**Comment by minad** [comment, marked as answer]
You can try to use `add-hook` instead, which adds to the global value of the variable. See: https://github.com/minad/cape/commit/826701cfc36dcc429ecb7174921be2085d04c1b9

**Reply by garyo** [reply]
Haha, I guess that would work even though `capf` is not actually a hook -- relying on `add-hook`'s behavior of always setting the global value. But it's simple to read, and it's also nice that it does nothing if the function is already present.
Thanks!

**Reply by minad** [reply]
Actually Capfs are hooks. They are invoked via `run-hook-wrapped`. I'd say there is really no distinction between functions and hooks. Hooks usually don't take arguments and they are all executed without stopping after some hook. But not even this holds in general, given `run-hook-with-args-until-success`.

**Reply by minad** [reply]
Btw, the discussion can stay open since it may help others and is more visible that way.
